field.go

```diff
@@ -187,14 +187,18 @@ func Error(err error) zapcore.Field {
 // extremely expensive (relatively speaking); this function both makes an
 // allocation and takes ~10 microseconds.
 func Stack(key string) zapcore.Field {
+	return stack(key, 4)
+}
+
+// needed for testing
+func stack(key string, skip int) zapcore.Field {
 	// Try to avoid allocating a buffer.
 	buf := bufferpool.Get()
-	bs := buf.Bytes()
 	// Returning the stacktrace as a string costs an allocation, but saves us
 	// from expanding the zapcore.Field union struct to include a byte slice. Since
 	// taking a stacktrace is already so expensive (~10us), the extra allocation
 	// is okay.
-	field := String(key, takeStacktrace(bs[:cap(bs)], false))
+	field := String(key, takeStacktrace(buf, skip))
```

Comment:
I think, string allocation for stack is large enough, and it could be fixed.
What about `BufferType` `FieldType`, with `*bufferpool.Buffer` in `Field.Interface` (don't allocate), that on AddTo calls `ObjectEncoder.AddBinary(bufferBytes)`, and then returns buffer to pool?
But in that case we should document, on `ObjectEccoder.AddBinary` that after `AddBinary` bytes under slice can be modified, so `ObjectEncoder` should not retain reference for it. It's reasonable constraint, `io.Writer` have same.  In zap only *MapObjectEncoder retains references for binary slices.